app/packages/core/src/components/Grid/useExpandSample.ts

```diff
@@ -1,47 +1,112 @@
-import { useCallback } from "react";
+import { Snapshot, useRecoilCallback } from "recoil";
 
-import { FlashlightConfig } from "@fiftyone/flashlight";
+import { ItemIndexMap } from "@fiftyone/flashlight/src/state";
 import * as fos from "@fiftyone/state";
+import { getFetchFunction } from "@fiftyone/utilities";
+
+import { pageParameters } from "./recoil";
+
+interface SamplesResponse {
+  results: Array<{
+    sample: fos.AppSample & { frame_number?: number };
+    aspect_ratio: number;
+    frame_rate?: number;
+    urls: Array<{ field: string; url: string }>;
+    thumbnails_only: boolean;
+  }>;
+}
+
+/**
+ * Returns a full sample by id, querying from the backend if necessary.
+ *
+ * Throws if the provided sampleId does not exist in the store.
+ */
+async function getFullSample<T extends fos.Lookers>(
+  store: fos.LookerStore<T>,
+  snapshot: Snapshot,
+  sampleId: string
+): Promise<fos.SampleData | null> {
+  const existingSample = store.samples.get(sampleId);
+
+  if (!existingSample) {
+    return null;
+  }
+
+  if (!existingSample.thumbnailsOnly) {
+    return existingSample;
+  } else {
+    // Query for the full sample from the backend
+    const { zoom, ...params } = await snapshot.getPromise(
+      pageParameters(/* modal= */ true)
+    );
+
+    const { results } = (await getFetchFunction()("POST", "/samples", {
+      ...params,
+      sample_id: sampleId,
+    })) as SamplesResponse;
+    const result = results[0];
+
+    const newSample: fos.SampleData = {
+      sample: result.sample,
+      aspectRatio: result.aspect_ratio,
+      frameRate: result.frame_rate,
+      frameNumber: result.sample.frame_number,
+      urls: Object.fromEntries(
+        result.urls.map(({ field, url }) => [field, url])
+      ),
+      thumbnailsOnly: false,
+    };
+
+    // Save the new sample in the global store
+    store.samples.set(result.sample._id, newSample);
+    return newSample;
+  }
+}
 
 export default <T extends fos.Lookers>(store: fos.LookerStore<T>) => {
   const expandSample = fos.useExpandSample();
   const setSample = fos.useSetExpandedSample(false);
   const clear = fos.useClearModal();
 
-  return useCallback<
-    (
-      ...args: Parameters<NonNullable<FlashlightConfig<number>["onItemClick"]>>
-    ) => void
-  >(
-    (next, sampleId, itemIndexMap) => {
-      const clickedIndex = itemIndexMap[sampleId];
+  return useRecoilCallback(
+    ({ snapshot }) =>
```

Comment:
`snapshot` stores the state of all Recoil variables at the time (used for getting `pageParameters`)